src/charm.py

```diff
@@ -0,0 +1,265 @@
+        #     self,
+        #     dependency_model=OpensearchDasboardsDependencyModel(
+        #         **DEPENDENCIES  # pyright: ignore[reportGeneralTypeIssues]
+        #     ),
+        # )
+
+        # --- MANAGERS ---
+
+        self.tls_manager = TLSManager(
+            state=self.state, workload=self.workload, substrate=SUBSTRATE
+        )
+        self.config_manager = ConfigManager(
+            state=self.state, workload=self.workload, substrate=SUBSTRATE, config=self.config
+        )
+
+        # --- LIB EVENT HANDLERS ---
+
+        self.restart = RollingOpsManager(self, relation="restart", callback=self._restart)
+
+        # --- CORE EVENTS ---
+
+        self.framework.observe(getattr(self.on, "install"), self._on_install)
+
+        self.framework.observe(getattr(self.on, "start"), self._manual_restart)
+
+        self.framework.observe(
+            getattr(self.on, "update_status"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, "leader_elected"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, "config_changed"), self._on_cluster_relation_changed
+        )
+
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_changed"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_joined"), self._on_cluster_relation_changed
+        )
+        self.framework.observe(
+            getattr(self.on, f"{PEER}_relation_departed"), self._on_cluster_relation_changed
+        )
+
+        self.framework.observe(getattr(self.on, "secret_changed"), self._on_secret_changed)
+
+    # --- CORE EVENT HANDLERS ---
+
+    def _on_install(self, event: InstallEvent) -> None:
+        """Handler for the `on_install` event."""
+        self.unit.status = MaintenanceStatus("installing Opensearch Dashboards...")
+
+        install = self.workload.install()
+        if not install:
+            self.unit.status = BlockedStatus("unable to install Opensearch Dashboards")
+
+        # don't complete install until passwords set
+        if not self.state.peer_relation:
+            self.unit.status = WaitingStatus("waiting for peer relation")
+            event.defer()
+            return
+
+        if self.unit.is_leader() and not self.state.cluster.internal_user_credentials:
+            for user in CHARM_USERS:
+                self.state.cluster.update({f"{user}-password": self.workload.generate_password()})
+
+    def _on_cluster_relation_changed(self, event: EventBase) -> None:
+        """Generic handler for all 'something changed, update' events across all relations."""
+        # not all methods called
+        if not self.state.peer_relation:
+            return
+
+        # attempt startup of server
+        if not self.state.unit_server.started:
+            self.init_server()
```

Comment:
` self.state.unit_server.started` is set once, at the beginning. Line 135 should be executed at most once throughout the runtime of the unit.

The particular event handler we're in at this point indeed is the "one and only" event handler for the cluster. The call on `self.config_manager.config_changed()` (soon after the quoted code above) only re-writes the config in case there was a change.

A restart is ONLY triggered if there was a config change. And indeed, the Dashboard doesn't seem to take config changes into account without a restart